### PR TITLE
Add support for QNX 7.1

### PR DESCRIPTION
**Pull request by chachoi**
Add support for QNX 7.1.

Build instruction:
```bash
# source the QNX environment script
source <path-to-qnxsdp-env.sh>/qnxsdp-env.sh

# Create a workspace
mkdir ~/vsomeip_workspace
WORKSPACE=~/vsomeip_workspace

# Build and install boost
cd $WORKSPACE
git clone https://gitlab.com/qnx/libs/boost.git && cd boost
git checkout QNX_7.1_v1.63.0
git submodule update --init --recursive
cd tools/build
git apply ../../boost_tools_build_qnx.patch
cd -
mkdir -p ./boost/utility
cp -r ./libs/utility/include/boost/utility/detail ./boost/utility
JLEVEL=8 make install

# Get googletest
cd $WORKSPACE
git clone https://gitlab.com/qnx/libs/googletest.git
export GTEST_ROOT=$WORKSPACE/googletest

# Build and install vsomeip
git clone https://github.com/chachoi/vsomeip.git && cd vsomeip/build_qnx

# Before building, look at $WORKSPACE/vsomeip/build_qnx/common.mk
# set TEST_IP_MASTER to the target ip address
# set TEST_IP_SLAVE to the host PC ip address for tests
JLEVEL=8 make install
```

[Test instruction](https://gitlab.com/qnx/libs/vsomeip/-/wikis/Test-vsomeip)

### CODE REVIEW

**Comment by chachoi** [comment]
@goncaloalmeida Hi, thanks for the feedback. I was able to remove a lot of unnecessary changes involving uid_t and gid_t.

File: interface/vsomeip/vsomeip_sec.h
Comment (fcmonteiro) [inline interface/vsomeip/vsomeip_sec.h:1]: Is this really needed?

File: implementation/utility/src/wrappers.cpp
[the remark this answers is not in the document]
Comment (chachoi) [reply implementation/utility/src/wrappers.cpp:1]: Yes, I moved all QNX related components to wrappers_qnx.cpp.

**Comment by goncaloalmeida** [comment]
Please update the PR and check the requested changes.
Thanks

**Comment by himanshu8921** [comment]
@chachoi SOME/IP support for QNX 7.1 by when it will be released ?

**Comment by nickchen8810** [comment]
@chachoi   I followed the exact steps in the guide and build the libs successfully.
But when I verified it on QOS222, the hello world demo of the vsomeip could not work properly.

# VSOMEIP_CONFIGURATION=/var/someip_test/helloworld-local.json VSOMEIP_APPLICATION_NAME=hello_world_service ./hello_world_service
1970-01-01 00:01:16.762084 [info] Using configuration file: "/var/someip_test/helloworld-local.json".
1970-01-01 00:01:16.762084 [info] Parsed vsomeip configuration in 2ms
1970-01-01 00:01:16.762084 [info] Configuration module loaded.
1970-01-01 00:01:16.762084 [info] Initializing vsomeip (3.3.8) application "hello_world_service".
1970-01-01 00:01:16.762084 [info] Instantiating routing manager [Host].
1970-01-01 00:01:16.762084 [info] create_routing_root: Routing root @ /var/vsomeip-0
1970-01-01 00:01:16.762084 [error] local_uds_server_endpoint_impl: bind failed (Permission denied)
1970-01-01 00:01:16.762094 [error] local_uds_server_endpoint_impl: listen failed (Invalid argument)

**Comment by goncaloalmeida** [comment]
@chachoi we are testing this PR internally.
Could you please provide an answer to @nickchen8810 problem?
thanks!
cc @DiogoPedrozza

**Comment by chachoi** [comment]
In my latest push, I changed `#ifdef __QNXNTO__` to `#ifdef __QNX__` because `__QNXNTO__` is deprecated. I also solved a merge conflict in `implementation/endpoints/src/endpoint_impl.cpp`.

File: implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp [hunk 1]
```diff
@@ -72,15 +98,29 @@ signed_size_type recv(socket_type s, buf* bufs, size_t count,
...
 	     cmsg = CMSG_NXTHDR(&msg, cmsg))
 	{
+#ifdef __QNX__
+	  if (cmsg->cmsg_level != SOL_SOCKET || cmsg->cmsg_type != SCM_CREDS
+	      || cmsg->cmsg_len != CMSG_LEN(sizeof(UCRED_T)))
+	    continue;
+#else
 	  if (cmsg->cmsg_level != SOL_SOCKET || cmsg->cmsg_type != SCM_CREDENTIALS
-	      || cmsg->cmsg_len != CMSG_LEN(sizeof(struct ucred)))
+	      || cmsg->cmsg_len != CMSG_LEN(sizeof(UCRED_T)))
```
Comment (DiogoPedrozza) [inline implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp:107]: Why this change? UCRED_T seems to only be defined within __QNX__ environments.

File: implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp [hunk 2]
```diff
@@ -122,13 +166,25 @@ signed_size_type recvfrom(socket_type s, buf* bufs, size_t count,
 
   union {
     struct cmsghdr cmh;
-    char   control[CMSG_SPACE(sizeof(struct ucred))];
+#ifdef __QNX__
+    char   control[512];
+#else
+    char   control[CMSG_SPACE(sizeof(UCRED_T))];
```
Comment (DiogoPedrozza) [inline implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp:172]: Another usage of UCRED_T outside of __QNX__ ifdef.

File: implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp [hunk 3]
```diff
@@ -143,15 +199,29 @@ signed_size_type recvfrom(socket_type s, buf* bufs, size_t count,
...
 	     cmsg = CMSG_NXTHDR(&msg, cmsg))
 	{
+#ifdef __QNX__
+	  if (cmsg->cmsg_level != SOL_SOCKET || cmsg->cmsg_type != SCM_CREDS
+	      || cmsg->cmsg_len != CMSG_LEN(sizeof(UCRED_T)))
+	    continue;
+#else
 	  if (cmsg->cmsg_level != SOL_SOCKET || cmsg->cmsg_type != SCM_CREDENTIALS
-	      || cmsg->cmsg_len != CMSG_LEN(sizeof(struct ucred)))
+	      || cmsg->cmsg_len != CMSG_LEN(sizeof(UCRED_T)))
```
Comment (DiogoPedrozza) [inline implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp:208]: Another one.

**Comment by Alwin-Bnglr** [comment]
@chachoi 
hello_world_service and hello_world_client using helloworld-local.json is with the service-discovery disabled.
Have you tried running any example with service-discovery enabled ?

**Comment by chachoi** [comment]
> @chachoi hello_world_service and hello_world_client using helloworld-local.json is with the service-discovery disabled. Have you tried running any example with service-discovery enabled ?

I tried hello_world_client and hello_world_service with service-discovery set to true, and it worked.

hello_world_service:
```console
# hello_world_service 
1970-01-01 00:40:56.565384 [info] Using configuration file: "/helloworld-local.json".
1970-01-01 00:40:56.565464 [info] Parsed vsomeip configuration in 8ms
1970-01-01 00:40:56.565524 [info] Configuration module loaded.
1970-01-01 00:40:56.565574 [info] Initializing vsomeip (3.3.8) application "hello_world_service".
1970-01-01 00:40:56.565664 [info] Instantiating routing manager [Host].
1970-01-01 00:40:56.565734 [info] create_routing_root: Routing root @ /var/vsomeip-0
1970-01-01 00:40:56.565814 [info] Service Discovery enabled. Trying to load module.
1970-01-01 00:40:56.566184 [info] Service Discovery module loaded.
1970-01-01 00:40:56.566244 [info] Application(hello_world_service, 4444) is initialized (11, 100).
1970-01-01 00:40:56.566324 [info] Starting vsomeip application "hello_world_service" (4444) using 2 threads I/O nice 255
1970-01-01 00:40:56.566434 [info] main dispatch thread id from application: 4444 (hello_world_service) is: 4
1970-01-01 00:40:56.566434 [info] shutdown thread id from application: 4444 (hello_world_service) is: 5
1970-01-01 00:40:56.566434 [error] udp_server_endpoint_impl: bind failed (Can't assign requested address)
1970-01-01 00:40:56.566714 [error] udp_server_endpoint_impl: set IPv4 outbound interface option failed (Can't assign requested add)
1970-01-01 00:40:56.566834 [warning] udp_server_endpoint_impl: couldn't set SO_RCVBUF: No buffer space available to: 1703936 local0
1970-01-01 00:40:56.566964 [info] udp_server_endpoint_impl: SO_RCVBUF is: 41600 (1703936) local port:30490
1970-01-01 00:40:56.567054 [debug] Joining to multicast group 224.224.224.0 from 134.86.56.94
1970-01-01 00:40:56.567144 [info] SOME/IP routing ready.
1970-01-01 00:40:56.567144 [warning] udp_server_endpoint_impl<multicast>: couldn't set SO_RCVBUF: No buffer space available to: 170
1970-01-01 00:40:56.567194 [info] Watchdog is disabled!
1970-01-01 00:40:56.567324 [info] udp_server_endpoint_impl<multicast>: SO_RCVBUF is: 41600 (1703936) local port:30490
1970-01-01 00:40:56.567374 [info] OFFER(4444): [1111.2222:0.0] (true)
1970-01-01 00:40:56.567374 [info] io thread id from application: 4444 (hello_world_service) is: 7
1970-01-01 00:40:56.567374 [info] io thread id from application: 4444 (hello_world_service) is: 1
1970-01-01 00:40:56.567544 [error] udp_server_endpoint_impl: bind failed (Can't assign requested address)
1970-01-01 00:40:56.567634 [info] vSomeIP 3.3.8 | (default)
1970-01-01 00:40:56.567814 [error] udp_server_endpoint_impl: set IPv4 outbound interface option failed (Can't assign requested add)
1970-01-01 00:40:56.567984 [warning] udp_server_endpoint_impl: couldn't set SO_RCVBUF: No buffer space available to: 1703936 local9
1970-01-01 00:40:56.568114 [info] udp_server_endpoint_impl: SO_RCVBUF is: 41600 (1703936) local port:30509
1970-01-01 00:40:56.568214 [info] create_local_server: Listening @ /var/vsomeip-4444
1970-01-01 00:41:06.667874 [info] vSomeIP 3.3.8 | (default)
1970-01-01 00:41:07.671324 [info] Application/Client 5555 is registering.
1970-01-01 00:41:07.671384 [info] Client [4444] is connecting to [5555] at /var/vsomeip-5555
1970-01-01 00:41:07.671484 [info] REGISTERED_ACK(5555)
1970-01-01 00:41:07.671584 [info] REQUEST(5555): [1111.2222:255.4294967295]
1970-01-01 00:41:07.671664 [info] RELEASE(5555): [1111.2222]
1970-01-01 00:41:07.671714 [info] Application/Client 5555 is deregistering.
1970-01-01 00:41:07.671794 [info] local_uds_client_endpoint_impl::receive_cbk Error: End of file
1970-01-01 00:41:07.671894 [info] Client [4444] is closing connection to [5555]
1970-01-01 00:41:12.721664 [info] STOP OFFER(4444): [1111.2222:0.0] (true)
1970-01-01 00:41:12.721724 [info] Stopping vsomeip application "hello_world_service" (4444).
```

hello_world_client:
```console
1970-01-01 00:41:07.671284 [info] Using configuration file: "/helloworld-local.json".
1970-01-01 00:41:07.671294 [info] Parsed vsomeip configuration in 8ms
1970-01-01 00:41:07.671294 [info] Configuration module loaded.
1970-01-01 00:41:07.671294 [info] Initializing vsomeip (3.3.8) application "hello_world_client".
1970-01-01 00:41:07.671294 [info] Instantiating routing manager [Proxy].
1970-01-01 00:41:07.671294 [info] Client [5555] is connecting to [0] at /var/vsomeip-0
1970-01-01 00:41:07.671294 [info] Application(hello_world_client, 5555) is initialized (11, 100).
1970-01-01 00:41:07.671294 [info] Starting vsomeip application "hello_world_client" (5555) using 2 threads I/O nice 255
1970-01-01 00:41:07.671304 [info] main dispatch thread id from application: 5555 (hello_world_client) is: 2
1970-01-01 00:41:07.671304 [info] shutdown thread id from application: 5555 (hello_world_client) is: 3
1970-01-01 00:41:07.671304 [info] io thread id from application: 5555 (hello_world_client) is: 4
1970-01-01 00:41:07.671304 [info] io thread id from application: 5555 (hello_world_client) is: 1
1970-01-01 00:41:07.671314 [info] create_local_server: Listening @ /var/vsomeip-5555
1970-01-01 00:41:07.671324 [info] Client 5555 (hello_world_client) successfully connected to routing  ~> registering..
1970-01-01 00:41:07.671324 [info] Registering to routing manager @ vsomeip-0
1970-01-01 00:41:07.671474 [info] Application/Client 5555 (hello_world_client) is registered.
Sending: World
1970-01-01 00:41:07.671644 [info] ON_AVAILABLE(5555): [1111.2222:0.0]
1970-01-01 00:41:07.671644 [info] Client [5555] is connecting to [4444] at /var/vsomeip-4444
Received: Hello World
1970-01-01 00:41:07.671664 [info] Stopping vsomeip application "hello_world_client" (5555).
1970-01-01 00:41:07.671784 [info] Application/Client 5555 (hello_world_client) is deregistered.
1970-01-01 00:41:07.671784 [info] local_uds_client_endpoint_impl::receive_cbk Error: Operation canceled
1970-01-01 00:41:07.671794 [info] local_uds_client_endpoint_impl::receive_cbk Error: Operation canceled
1970-01-01 00:41:07.671794 [info] Client [5555] is closing connection to [4444]
```

File: implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp (hunk 1 above)
Comment (chachoi) [reply implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp]: This is a mistake. Thanks for pointing it out.

File: implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp (hunk 2 above)
Comment (chachoi) [reply implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp]: Fixed.

File: implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp (hunk 3 above)
Comment (chachoi) [reply implementation/helper/boost/asio/detail/impl/socket_ops_ext_local.ipp]: Fixed.

**Comment by Alwin-Bnglr** [comment]
@chachoi 
Thanks for trying the example with service-discovery enabled. 
I am too able to replicate the same.

But once tried between two different hosts, I am not able to see the communication getting established.
1. Have you tried this PR with service/clients on different host devices ?
2. If yes, can you please mention the pre-requisites for that ? [eg : making the QNX device configured for receiving multicast messages ]

**Review by DiogoPedrozza — APPROVED** [review]

File: build_qnx/common.mk [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+#information for shared objects that is used by the QNX build
+#infrastructure to embed metadata in the .so files, for example
+#data and time, version number, description, etc. Metadata can
+#be retrieved on the target by typing 'use -i <path to vsomeip .so file>'.
+#this is optional: setting GENERATE_PINFO_FILES to FALSE will disable
+#the insertion of metadata in .so files.
+ifeq ($(GENERATE_PINFO_FILES), TRUE)
+#the following rules are called by the cmake generated makefiles,
+#in order to generate the .pinfo files for the shared libraries
+%.so.3.3.8:
```
Comment (goncaloalmeida) [inline build_qnx/common.mk:96]: why hardcoded version?
We plan to merge this PR today or tomorrow, but we can not merge with hardcoded version (if we update vsomeip version, the so libraries will change the name also).

**Review by goncaloalmeida — CHANGES_REQUESTED** [review]

File: build_qnx/common.mk (hunk 4 above)
Comment (goncaloalmeida) [reply build_qnx/common.mk:96]: @chachoi

Comment (chachoi) [reply build_qnx/common.mk]: Thanks for the feedback. I fixed it so that the version information is now parsed from the CMakeLists.txt.

**Review by goncaloalmeida — APPROVED** [review]

**Review by DiogoPedrozza — APPROVED** [review]
